File: two pointer/typescript/DutchNationalFlagProblem.ts
/**
 Do not return anything, modify nums in-place instead.
 */
function sortColors(nums: number[]): number[] {

    let left = 0;
    let right = nums.length - 1;
    let i = 0;

    while (i <= right) {
        let num = nums[i];

        if (num === 0) {
            nums[i] = nums[left];
            nums[left] = num;
            i++;
            left++;
        }
        else if (num === 2) {
            nums[i] = nums[right];
            nums[right] = num;
            right--;
        }
        else
            i++;

    }

    return nums;

};

console.log(sortColors([2, 0, 2, 1, 1, 0]));
console.log(sortColors([2, 0, 1]));
console.log(sortColors([0, 0, 2, 1, 1, 2]));
console.log(sortColors([0, 0, 0, 0]));

File: two pointer/java/DutchNationalFlagProblem.java
import java.util.Arrays;

public class DutchNationalFlagProblem {
    public static void main(String[] args) {
        System.out.println(Arrays.toString(sortColors(new int[] { 2, 0, 2, 1, 1, 0 })));
        System.out.println(Arrays.toString(sortColors(new int[] { 2, 0, 1 })));
        System.out.println(Arrays.toString(sortColors(new int[] { 0, 0, 2, 1, 1, 2 })));
        System.out.println(Arrays.toString(sortColors(new int[] { 0, 0, 0, 0 })));
    }

    public static int[] sortColors(int[] nums) {

        int i = 0;
        int left = 0;
        int right = nums.length - 1;

        while (i <= right) {
            int num = nums[i];

            if (num == 0) {
                nums[i] = nums[left];
                nums[left] = num;
                left++;
                i++;
            } else if (num == 2) {
                nums[i] = nums[right];
                nums[right] = num;
                right--;
            } else
                i++;
        }

        return nums;

    }
}
